# Examen Ordinario Econometría 2020-2021. ADE+Ingenierías
# Tipo A
## 22 de Enero de 2021


### Instrucciones

+ La duración del examen de 1.5 horas.
+ Cada pregunta se responderá en el recuadro bajo la pregunta, bien en modo sólo texto o imagen (Preguntas 1 a 5) o bien en modo códigos de Python + texto respuesta en base a los resultados obtenidos (Preguntas 6 a 10).
+ Los datos necesarios está disponibles bien en el, propio ejercicio, o en la librería wooldridge o en los ficheros de excel *datos8.xls* y *datos9.xls*) disponibles en github (las órdenes orientativas para leer los ficheros se encuentran descritas bajo cada uno de esos ejercicios).
+ No se evaluarán respuestas en las que únicamente haya códigos. Los resultados deberán servir para responder a las preguntas.
+ Las puntuaciones de cada ejercicio se encuentran en el propio enunciado.
+ El fichero final del examen se enviará a vía PRADO en la Tarea habilitada para el examen.


### APELLIDOS,  NOMBRE: 
### DNI: 

### Pregunta 1: (1 punto)

Demuestra que en un modelo, $y_i = \beta_0 + \beta_1 x_{1i} + \cdots + \beta_{k} x_{ki} + u_i$ ($i=1, \ldots, n$) estimado por Mínimos cuadrados Ordinarios, se verifica que
$$
\sum_{i=1}^n (y_i-\bar y)^2 = \sum_{i=1}^n (y_i - \widehat{y}_i)^2+ \sum_{i=1}^n (\widehat{y}_i-\bar y)^2.
$$
¿Qué consecuencias tiene este resultado en el análisis del modelo?

- Respuesta: 

  -   Esta formula se basa en el modelo SCT = SCR + SCE, esto nos permitira sacar la sumatoria de los residuos, explicada, totales... 
  -   Podemos sacar por ejemplo, el r-squared, podemos hacer test experimentales, la estimacion de la varianza, el de anova, test de significacion...


### Pregunta 2: (1 punto)

Considera el modelo $Y_i = \beta_0 + \beta_1 X_{1i} + \beta_2 X_{2i} + u_i$ ($i=1,2,\ldots, 50$). Se sabe que $X_{1i}= \dfrac{1}{4} X_{2i}$. ¿Qué coeficientes de este modelo pueden estimarse directamente y cuáles no? Obtener las combinaciones lineales de coeficientes que son estimables por Mínimos Cuadrados Ordinarios.

- Respuesta: 
  
  - Esto nos indica que se presenta una multicolinealidad entre X1 y X2, lo que implica que a la hora de sacar las betas no se podrian estimar de manera correcta ya que la multiplacion de (Xt*X)-1 * Xty, no nos daria resultados fiables al ser una combinacion lineal de otras variables.

  - Se presenta una multicolienalidad perfecta, porque literalmente el X1 es combinacion lineal de X2.

  - Beta_hat = (Xt*X)-1 * Xty 
  - Yi = Beta[0] + Beta[1]* (1/4)x2 + Beta[2]*x2 
  - --> Yi = Beta[0] + x2(Beta[1] * 1/4 + Beta[2]) 

### Pregunta 3: (0.75 puntos)

en el modelo $Y_i= \beta_0 + \beta_1 X_i + u_i$, $i=1, \ldots, n$, cuyas perturbaciones no están autocorreladas, pero son tales que $var(u_i)=\frac{\sigma^2}{\sqrt{X_i}}$, ¿Cómo obtendrías el estimador más adecuado de $\beta_1$? Comprobar que se ha corregido el posible incumplimiento en la shipótesis básicas.

Respuesta:
 
 - En primer lugar podemos ver que hay heterocedasticidad ya que podemos ver que nuestra varianza esta dependiendo del propio individuo, sin haber autocorrelacion ya que las perturbaciones no estan autocorreladas.
  
- Con minimos cuadrados ponderados podemos sacar: 
    $\Omega$ se convierte en la matriz varianzas covarianzas diagonalizada con w<sub>i</sub>, donde w<sub>i</sub> = $\frac{\sigma^2}{\sqrt{X_i}}$, con lo que $var(u_i) = {\sigma^2} * {w_i}$, haciendo la inversa de $\Omega$ y considerando P nuestra nueva matriz con 1/${\sqrt{w_i}}$.

- Luego despues de multiplicar P por y, por X y por u, podemos deducir que Var(${u_i^*}$) = Var($\frac{u_i}{\sqrt{w_i}}$) = $\frac{1}{w_i}$ Var(${w_i}$) = $\frac{1}{w_i}$ ${\sigma^2}$ * ${w_i}$ = ${\sigma^2}$.

- Con lo que podemos determinar que ya no habria heterocedasticidad
- Las demostraciones son correctas pag 14 Tema 5

### Pregunta 4: (0.5 puntos) Esta hecha ez

En el caso de tener que decidir qu\'e modelo es el mejor en el siguiente caso:
        
            Modelo 1: & & \mathbf{Y} = \beta_{1} + \beta_{2} \cdot \mathbf{X}_{2} + \beta_{3} \cdot \mathbf{X}_{3} + \alpha_{4} \cdot \mathbf{X}_{4} + \mathbf{u}. \\
            Modelo 2: & & \mathbf{Y} = \alpha_{1} + \alpha_{2} \cdot \mathbf{X}_{2} + \alpha_{3} \cdot \mathbf{X}_{3} + \mathbf{v}.   
        
¿Podemos afirmar que el $R^2$ de alguno de los modelos es mejor que el del otro? ¿Cómo podemos decidir cuál es el mejor modelo? Justificar su respuesta.

- Respuesta:

- A la hora de comprara dos modelos con la misma variable endogena podemos utilizar los R^2 para comparar los dos modelos, hay que tener en cuenta que los dos modelos son iguales, solo que el segundo modelo tiene una variable mas, lo que podria darnos casos de sobreestimacion del modelo.
  
- Por eso podriamos juzgarlos con el R^2 ajustado que penaliza el hecho de utilizar mas variables, ya que esto va a hacer que el R^2 vaya a ser mejor siempre pero el hecho de añadir variables por añadir podria dar ciertos problemas, de sobrestimacion del modelo, cambios en las significaciones individuales, incremento de multicolinealidad,etc...

- En caso de que la variable endogena no fuese la misma, podriamos utilizar otros criterios de seleccion como el AIC (Akaike) y BIC (Schwarz).
  
- El modelo mas correcto seria el que tuviese mayor numero de variables, porque es el mejor lo explica porque va a tener un r-squared mayor.
- Estimamos con los minimos cuadrados ordinario y comprobamos.
- No necesariamente funciona con el r-squared, porque a veces se puede producir una sobreestimacion(que se utilicen un mayor numero de variables de las necesarias).
- Lo correcto es utilizar el r-squared ajustado, porque este se ajusta al numero de variables que hay.


### Pregunta 5: (0.75 puntos)

Consideremos un modelo económico en el que se desea explicar, utilizando una base de datos de 109 viviendas, el precio de la vivienda ($Y$) en función de la zona donde se encuentra la vivienda ($X_1$), sus metros cuadrados ($X_2$) y el número de habitaciones ($X_3$). Se han realizado las regresiones auxiliares:
$$
X_{1i} = \beta_0 + \beta_1 X_{2i} + \beta X_{3i} + u_i, i=1, \ldots, 109,
$$
$$
X_{2i} = \beta_0 + \beta_1 X_{1i} + \beta X_{3i} + u_i, i=1, \ldots, 109,
$$
$$
X_{3i} = \beta_0 + \beta_1 X_{1i} + \beta X_{2i} + u_i, i=1, \ldots, 109,
$$
obteniéndose como $R^2$ los valores $0.85$, $0.93$ y $0.75$. ¿Qué podemos decir sobre el modelo analizado en base a esta información?

- Respuesta: Puedes estar sobreestimando ya que realmente la variable X2 puede ser explicada por las variables X1 y X3, porque puede ser que se presente un problema de multicolinealidad y lo mismo para los otros casos?
- Para calcular el vif de cada una de las variables es:
- 1/1-Ri^2 --> Asi comprobamos si tenemos problemas de multicolinealidad
- Nuestro vif primero es 6.6 --> No hay problema
- Nuestro segundo vif es 14 --> Tenemos problema de multicolinealidad (no grave)
- Nuestro tercer vif es 4 --> No hay problema

### Pregunta 6: (1.5 puntos)

Se desea analizar el modelo $Y_ t = \beta_0 + \beta_1 X_{1t} + \beta_2 X_{2t} + u_t$, con 
        $$(\mathbf{X}^{t} \mathbf{X}) =  \begin{pmatrix} 14 & 7 & 14\\ 7 & 4.5 & 7\\ 14 & 7 & 15 \end{pmatrix}, \mathbf{X}^{t} \mathbf{y} = 
            \begin{pmatrix}
                10\\6\\12
            \end{pmatrix}, \mathbf{y}^t \mathbf{y} =14$$
            
Se pide responder de forma razonada las siguientes cuestiones:
1. Obtener la estimación mínimo cuadrática de los coeficientes del modelo.
1. Verificar la signifaicación del modelo.
1. Calcular la predicción puntual para el valor individual, así como el intervalo de predicción para este valor al $95\%$ sabiendo que $x_1=5$ y $x_2=7$.

In [12]:
#Cálculos:

#Cálculos:
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
import matplotlib.pylab as plt
from wooldridge import *
import statsmodels.stats.outliers_influence as oi #Para el test RESET de Ramsey
import statsmodels.stats.diagnostic as diagn #Para el test de Harvey-Collier
import pandas as pd
from scipy import stats

X_trasp_X = np.array([[14,7,14],[7,4.5,7],[14,7,15]])

X_trasp_X_inversa = np.linalg.inv(X_trasp_X)

X_trasp_y = np.array([[10], [6], [12]])
y_trasp_y = 14

Beta_hat = np.dot(X_trasp_X_inversa, X_trasp_y)
print("Nuestro beta hat: ", Beta_hat)

n = X_trasp_X[0][0]
k = 2
print("Numero de muestras: ", n)
print("Numero de variables: ", k)

scr = y_trasp_y - np.dot(np.transpose(Beta_hat), X_trasp_y)
sce = np.dot(np.transpose(Beta_hat), X_trasp_y) - np.dot(n, (X_trasp_y[0]/n)**2)

alpha = 0.05

F_exp = ((n-k)/(k-1))*(sce/scr)
F_teorica = stats.f.ppf((1-alpha), (k-1), (n-k))

print("\nF teorica: ", F_teorica)
print("F experimental: ", F_exp)
print("Como nuestra f-experimental es mayor que la f-teorica, se rechaza la hipotesis nula y decimos que tenemos significacion global")

#------------------------------------------
# Apartado 3

sigma_hat_sq = scr/(n-k)
matriz_varianza_covarianza = sigma_hat_sq * X_trasp_X_inversa

t_exp_1 = np.absolute(Beta_hat[1]/np.sqrt(matriz_varianza_covarianza[1,1]))
t_exp_2 = np.absolute(Beta_hat[2]/np.sqrt(matriz_varianza_covarianza[2,2]))
t_teorica = stats.t.ppf(1-(alpha/2), n-k)

print("\nPara nuestra beta_hat[1]: ")
print("T teorica: ", t_teorica)
print("T experimental: ", t_exp_1)


print("\nPara nuestra beta_hat[2]: ")
print("T teorica: ", t_teorica)
print("T experimental: ", t_exp_2)

Y = Beta_hat[0] + Beta_hat[1]*5 + Beta_hat[2]*7
print("\nNuestra prediccion es: ", Y)



Nuestro beta hat:  [[-1.78571429]
 [ 1.        ]
 [ 2.        ]]
Numero de muestras:  14.0
Numero de variables:  2

F teorica:  4.747225346722511
F experimental:  [[32.30769231]]
Como nuestra f-experimental es mayor que la f-teorica, se rechaza la hipotesis nula y decimos que tenemos significacion global

Para nuestra beta_hat[1]: 
T teorica:  2.1788128296634177
T experimental:  [2.54195564]

Para nuestra beta_hat[2]: 
T teorica:  2.1788128296634177
T experimental:  [5.08391127]

Nuestra prediccion es:  [17.21428571]


- Respuesta:


### Pregunta 7: (1.5 puntos)

Se considera el modelo econométrico $Y_t = \beta_0 X_{t}^{\beta_1} Z_t^{\beta_2} {\rm exp}({u_i})$ $i=1899, \ldots, 1922$, donde $Y$ es la producción total de un bien, $X$ es el stock de capital, $Z$ es el número de empleados para producir el bien y $u_t$ es la perturbación aleatoria durante los años $1899$ a $1922$.
1. Justifica cómo se puede analizar el modelo utilizando Mínimos Cuadrados Ordinarios.
1. Estima el modelo para los siguientes e interpreta los coeficientes obtenidos:

$X: [100, 107, 114, 122, 131, 138, 149, 163, 176, 185, 198, 208, 216, 226, 236, 244, 266, 298, 335, 366, 387, 407, 417, 431]$

$Z: [100, 105, 110, 118, 123, 116, 125, 133, 138, 121, 140, 144, 145, 152, 154, 149, 154, 182, 196, 200, 193, 193, 147, 161]$

$Y: [100, 101, 112, 122, 124, 122, 143, 152, 151, 126, 155, 159, 153, 177, 184, 169, 189, 225, 227, 223, 218, 231, 179, 240]$

1. Comprueba si $\beta_1+\beta_2=1$ e interpreta el resultado obtenido.

In [62]:
#Cálculos:

Y = np.log(np.array([100, 101, 112, 122, 124, 122, 143, 152, 151, 126, 155, 159, 153, 177, 184, 169, 189, 225, 227, 223, 218, 231, 179, 240]))
XZ = np.log(np.array([ [100, 105, 110, 118, 123, 116, 125, 133, 138, 121, 140, 144, 145, 152, 154, 149, 154, 182, 196, 200, 193, 193, 147, 161], [100, 107, 114, 122, 131, 138, 149, 163, 176, 185, 198, 208, 216, 226, 236, 244, 266, 298, 335, 366, 387, 407, 417, 431]]))
XZ = np.transpose(XZ)

mco7 = sm.OLS(Y, sm.add_constant(XZ)).fit()
mco7.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.957
Model:                            OLS   Adj. R-squared:                  0.953
Method:                 Least Squares   F-statistic:                     236.1
Date:                Sun, 19 Dec 2021   Prob (F-statistic):           4.04e-15
Time:                        12:42:07   Log-Likelihood:                 35.826
No. Observations:                  24   AIC:                            -65.65
Df Residuals:                      21   BIC:                            -62.12
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.1773      0.434     -0.408      0.687      -1.080       0.726
x1             0.8073      0.145      5.565      0.000       0.506       1.109
x2             0.2331      0.064      3.668      0.001       0.101       0.365
==============================================================================
Omnibus:                        2.133   Durbin-Watson:                   1.523
Prob(Omnibus):                  0.344   Jarque-Bera (JB):                1.361
Skew:                           0.583   Prob(JB):                        0.506
Kurtosis:                       2.992   Cond. No.                         285.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- Respuesta:


### Pregunta 8: (1 punto)

Se pretende estudiar el efecto del crecimiento de los salarios en la inflación. Para ello se construye el modelo $Y_t = \beta_0 + \beta_1 X_t^2 + u_t$ con $t$ indicando los cuatrimestres desde 1970 a 2010, y donde $Y$ es la inflación y $X$ es el crecimiento porcentual de los salarios en ese periodo.
Estima el modelo lineal utilizando los datos de *datos1.xls* y estudia la existencia de autocorrelación en éste. En caso afirmativo, aplica alguna estrategia para corregirla y comprueba si ha concluido con éxito.

In [26]:
#Cálculos:

import pandas as pd

datos8 = pd.read_excel ("datos1.xls")

Y = datos8["inflacion"]
X_sqr = datos8["CrecimientoSalarios"]**2

mco8 = sm.OLS(Y, sm.add_constant(X_sqr)).fit()

mco8.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              inflacion   R-squared:                       0.209
Model:                            OLS   Adj. R-squared:                  0.204
Method:                 Least Squares   F-statistic:                     41.71
Date:                Sun, 19 Dec 2021   Prob (F-statistic):           1.24e-09
Time:                        11:23:40   Log-Likelihood:                -176.99
No. Observations:                 160   AIC:                             358.0
Df Residuals:                     158   BIC:                             364.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                   0.7931      0.075     10.577      0.000       0.645       0.941
CrecimientoSalarios     0.1881      0.029      6.458      0.000       0.131       0.246
==============================================================================
Omnibus:                       22.152   Durbin-Watson:                   0.851
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              101.899
Skew:                          -0.230   Prob(JB):                     7.46e-23
Kurtosis:                       6.883   Cond. No.                         3.54
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [27]:
#Cálculos:
from statsmodels.stats.stattools import durbin_watson #Si DW es 0, correlacion positiva, si es 4, correlacion negativa y si es 2 es incorrelaccionado, si no es alguno de esos valores me fijo en las tablas

DW = durbin_watson(mco8.resid) #con el DW 0.066 tengo un problema de correlacion positiva
print("DW: ", DW, "--> Problema de autocorrelacion")

DW:  0.8513677540550563 --> Problema de autocorrelacion


In [28]:
rho= 1 - DW/2 # ro estimado inicial

print (rho) # rho inicial
mco_autocorr=sm.GLSAR(Y, sm.add_constant(X_sqr), rho=rho) # minimos cuadrados generalizados
res = mco_autocorr.iterative_fit(maxiter=1000, rtol = 10**(-10)) # iteraciones y tolerancia, la tolerancia es si entre un rho y otro hay muy poca diferencia paro (10^-10)

print ('Iterations used = %d Converged %s' % (res.iter, res.converged) )
print ('Rho = ', mco_autocorr.rho)
res.summary()

0.5743161229724718
Iterations used = 12 Converged True
Rho =  [0.71998043]


<class 'statsmodels.iolib.summary.Summary'>
"""
                           GLSAR Regression Results                           
==============================================================================
Dep. Variable:              inflacion   R-squared:                       0.010
Model:                          GLSAR   Adj. R-squared:                  0.004
Method:                 Least Squares   F-statistic:                     1.609
Date:                Sun, 19 Dec 2021   Prob (F-statistic):              0.206
Time:                        11:24:42   Log-Likelihood:                -131.82
No. Observations:                 159   AIC:                             267.6
Df Residuals:                     157   BIC:                             273.8
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                   1.0261      0.164      6.270      0.000       0.703       1.349
CrecimientoSalarios     0.0339      0.027      1.269      0.206      -0.019       0.087
==============================================================================
Omnibus:                       95.903   Durbin-Watson:                   2.232
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1108.290
Skew:                          -1.897   Prob(JB):                    2.18e-241
Kurtosis:                      15.365   Cond. No.                         6.35
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

DW --> 1.746 < 2.232 < 2.254 (4-DU)

- Respuesta:


### Pregunta 9: (1 punto)

La base de datos includida en "datos2.xls" contiene información acerca de los precios de 940 viviendas, asi como sus metros cuadrados (m2) y su antigüedad (antiguedad). Utliza el test de White para chequear si se cumple alguna de la hipotesis básicas del modelo lineal, y en caso afirmativo, describe la problemática y sus consecuencias.

In [31]:
#Cálculos:

import pandas as pd

datos9 = pd.read_excel ("datos2.xls")
datos9

Y = datos9["Precio"]
X = datos9[["antiguedad", "m2"]]

mco9 = sm.OLS(Y, sm.add_constant(X)).fit()

mco9.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Precio   R-squared:                       0.567
Model:                            OLS   Adj. R-squared:                  0.566
Method:                 Least Squares   F-statistic:                     612.6
Date:                Sun, 19 Dec 2021   Prob (F-statistic):          7.40e-171
Time:                        11:28:35   Log-Likelihood:                -10753.
No. Observations:                 940   AIC:                         2.151e+04
Df Residuals:                     937   BIC:                         2.153e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       5054.9160   3586.477      1.409      0.159   -1983.542    1.21e+04
antiguedad  -217.3899     35.077     -6.198      0.000    -286.228    -148.551
m2           737.0764     23.343     31.576      0.000     691.266     782.887
==============================================================================
Omnibus:                      340.045   Durbin-Watson:                   2.044
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2467.857
Skew:                           1.462   Prob(JB):                         0.00
Kurtosis:                      10.379   Cond. No.                         685.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [33]:
import statsmodels.stats.api as sms

W=sms.het_white(mco9.resid, mco9.model.exog)
print("Test de White: ")
print("Descripcion --> Lo mismo que los anteriores test")
print(W, "--> Tengo problema de heterocedasticidad, ya que mi p-valor < 0.05 \n")

Test de White: 
Descripcion --> Lo mismo que los anteriores test
(35.52054602613952, 1.1843741030391916e-06, 7.335974265120917, 9.3440399192803e-07) --> Tengo problema de heterocedasticidad, ya que mi p-valor < 0.05 



- Respuesta: Mis problemas son que las varianzas no son iguales para todos los residuos y van variando, y segun markov, podemos decir que las betas no son optimos.


### Pregunta 10: (1 punto)

En la base de datos "crime2" de Wooldridge se facilita información sobre la tasa de crimen (crmrte) en diversas ciudades en los años 1982 y 1987 (variable year). En particular, estamos intereaados en utilizar esta base de datos para  saber como afecta el desempleo en la tasa de delincuencia. 
1. Analiza el modelo lineal que consideres más adecuado para este fin y comenta los resultados obtenidos para toda la base de datos.
2. ¿Existen diferencias significativas entre los resultados obtenidos para 1982 y para los de 1987?

In [44]:
#Cálculos:

from wooldridge import *
datos10=dataWoo("crime2")

Y = datos10["crmrte"]
X = datos10[["unem", "crimes", "officers", "pop", "area"]]

mco10 = sm.OLS(Y, sm.add_constant(X)).fit()
mco10.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 crmrte   R-squared:                       0.696
Model:                            OLS   Adj. R-squared:                  0.678
Method:                 Least Squares   F-statistic:                     39.31
Date:                Sun, 19 Dec 2021   Prob (F-statistic):           8.00e-21
Time:                        11:36:04   Log-Likelihood:                -387.75
No. Observations:                  92   AIC:                             787.5
Df Residuals:                      86   BIC:                             802.6
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        107.3603      5.306     20.234      0.000      96.812     117.908
unem          -0.4157      0.553     -0.751      0.455      -1.516       0.684
crimes         0.0019      0.000     13.057      0.000       0.002       0.002
officers      -0.0070      0.006     -1.187      0.238      -0.019       0.005
pop           -0.0002   1.96e-05     -9.270      0.000      -0.000      -0.000
area          -0.0064      0.019     -0.336      0.738      -0.045       0.032
==============================================================================
Omnibus:                       10.084   Durbin-Watson:                   1.387
Prob(Omnibus):                  0.006   Jarque-Bera (JB):               20.080
Skew:                           0.303   Prob(JB):                     4.36e-05
Kurtosis:                       5.207   Cond. No.                     1.42e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.42e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [46]:
#Cálculos:

datos1982 = datos10[datos10["year"] == 82]

Y=datos1982["crmrte"]
X=datos1982[["unem", "crimes", "officers", "pop", "area"]]

mco10_1982 = sm.OLS(Y, sm.add_constant(X)).fit()

mco10_1982.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 crmrte   R-squared:                       0.657
Model:                            OLS   Adj. R-squared:                  0.614
Method:                 Least Squares   F-statistic:                     15.33
Date:                Sun, 19 Dec 2021   Prob (F-statistic):           2.06e-08
Time:                        11:37:17   Log-Likelihood:                -186.16
No. Observations:                  46   AIC:                             384.3
Df Residuals:                      40   BIC:                             395.3
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         94.8924      8.832     10.744      0.000      77.043     112.742
unem           0.7006      0.798      0.878      0.385      -0.912       2.313
crimes         0.0020      0.000      8.171      0.000       0.002       0.003
officers      -0.0143      0.010     -1.468      0.150      -0.034       0.005
pop           -0.0002   3.07e-05     -5.717      0.000      -0.000      -0.000
area          -0.0010      0.024     -0.043      0.966      -0.050       0.048
==============================================================================
Omnibus:                       10.729   Durbin-Watson:                   2.068
Prob(Omnibus):                  0.005   Jarque-Bera (JB):               14.858
Skew:                           0.682   Prob(JB):                     0.000594
Kurtosis:                       5.427   Cond. No.                     1.88e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.88e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [47]:
#Cálculos:

datos1987 = datos10[datos10["year"] == 87]

Y=datos1987["crmrte"]
X=datos1987[["unem", "crimes", "officers", "pop", "area"]]

mco10_1987 = sm.OLS(Y, sm.add_constant(X)).fit()

mco10_1987.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 crmrte   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.720
Method:                 Least Squares   F-statistic:                     24.11
Date:                Sun, 19 Dec 2021   Prob (F-statistic):           4.16e-11
Time:                        11:37:41   Log-Likelihood:                -196.06
No. Observations:                  46   AIC:                             404.1
Df Residuals:                      40   BIC:                             415.1
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        132.4900     11.660     11.363      0.000     108.924     156.055
unem          -4.9312      2.013     -2.449      0.019      -9.000      -0.862
crimes         0.0018      0.000      9.472      0.000       0.001       0.002
officers      -0.0012      0.008     -0.156      0.877      -0.017       0.015
pop           -0.0002   2.67e-05     -7.035      0.000      -0.000      -0.000
area           0.0146      0.031      0.468      0.642      -0.048       0.077
==============================================================================
Omnibus:                        2.942   Durbin-Watson:                   2.048
Prob(Omnibus):                  0.230   Jarque-Bera (JB):                2.203
Skew:                           0.143   Prob(JB):                        0.332
Kurtosis:                       4.033   Cond. No.                     2.07e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.07e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [49]:
scr1982=mco10_1982.ssr
scr1987=mco10_1987.ssr
scr = mco10.ssr
n = mco10.nobs

k=3

alpha=0.05
Fteo= stats.f.ppf(1-alpha,k+1,n-2*k-1)

#Test de chow
Fexp = ((n-(2*k)-2)/(k-1)) * ((scr-(scr1987+scr1982))/(scr1987+scr1982))

print("La F-teorico: ", Fteo)
print("La F-experimental: ", Fexp)

print("Como nuestra f-experimental es mas grande que f-teorico, rechazamos la hipotesis nula, por lo que podemos decir que no coincide las betas de 1982 y 1987")

La F-teorico:  2.479015469591249
La F-experimental:  4.288571331232776
Como nuestra f-experimental es mas grande que f-teorico, rechazamos la hipotesis nula, por lo que podemos decir que no coincide las betas de 1982 y 1987


- Respuesta:
